## **Social Media Sentiment Analysis**

The objective of this component is to analyze the sentiment of social media comments for a nonprofit organization.

I am Member 5 of the team, and my responsibility is to work on Hugging Face Transformers.

# **Mini plan:**
I developed a sentiment analysis component for the Social Media Sentiment Dashboard. I used Hugging Face Transformers to analyze social media comments/text and I tested the model on a sample dataset and evaluate its predictions. The final output provided a sentiment label and confidence score for integration into the dashboard.

# **Model Selection:**
**Selected Model:** cardiffnlp/twitter-roberta-base-sentiment-latest

I chose this model because this model is suitable for Twitter/social media texts sentiment analysis, this model contain 3 sentiment labels (Positive/Negative/Neutral), and important of all we didn't need to train model from scratch it is pre-trained.

# **Sample Dataset:**
I created this dataset to test the model with different text/comments. It includes all type of sentiments (Positive/Negative,Neutral). I included all types of comments to check that either model identifies what people are saying in their comments. I created these 5 sample dataset to test this model:

*   "You are doing great work."
*   "Well I don't trust this."
*   "Does this organization work on the charity given by people?"
*   "Yeah I can see how much work is done."
*   "I'm happy this type of people exist who work for other people."

Model will give result in label and score form.











# **Initial Model Testing:**
First we will test the model with 3 cases.


In [ ]:
!pip install -q transformers torch

In [ ]:
from transformers import pipeline

In [ ]:
sentiment_pipeline = pipeline(
    "sentiment-analysis",
    model="cardiffnlp/twitter-roberta-base-sentiment-latest"
)

config.json:   0%|          | 0.00/929 [00:00<?, ?B/s]

pytorch_model.bin: reconstructing file:   0%|          |  0.00B /  501MB            

pytorch_model.bin: downloading bytes:           |  0.00B            

model.safetensors: reconstructing file:   0%|          |  0.00B /  501MB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

[transformers] RobertaForSequenceClassification LOAD REPORT from: cardiffnlp/twitter-roberta-base-sentiment-latest
Key                         | Status     |  | 
----------------------------+------------+--+-
roberta.pooler.dense.weight | UNEXPECTED |  | 
roberta.pooler.dense.bias   | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.


vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

In [ ]:
result = sentiment_pipeline("You are doing great work.")
result

[{'label': 'positive', 'score': 0.9711529016494751}]

In [ ]:
result = sentiment_pipeline("Well I don't trust this.")
result

[{'label': 'negative', 'score': 0.8766955733299255}]

In [ ]:
result = sentiment_pipeline("Does this organization work on the charity given by people?")
result

[{'label': 'neutral', 'score': 0.932409942150116}]

# **Initial test results:**
I tested first three datasets on model, positive/negative/neutral were included to check that model is not only detecting positive and negative it can also detect neutral and other type of comments and good thing is that prediction matched the same expected sentiments, it also produced score for each prediction.

# **Test Cases:**
Now in this step i tested 10 more cases as it was mentioned in the task.
| Type                    | Approx. cases |
| ----------------------- | ------------: |
| Clearly Positive        |             3 |
| Clearly Negative        |             3 |
| Neutral                 |             2 |
| Challenging / Ambiguous |             2 |
| **Total**               |        **10** |


In [ ]:
test_comments = [
    "You are doing great work.",
    "Well I don't trust this.",
    "Does this organization work on the charity given by people?",
    "Yeah I can see how much work is done.",
    "I'm happy this type of people exist who work for other people.",
    "Thank you for supporting our community and helping those in need.",
    "I am disappointed with how this organization handled the donations.",
    "Where can I find information about your next campaign?",
    "Another campaign, but nothing ever seems to change.",
    "The volunteers have done an amazing job helping local families."
]

In [ ]:
results = sentiment_pipeline(test_comments)
results

[{'label': 'positive', 'score': 0.9711529016494751},
 {'label': 'negative', 'score': 0.8766955733299255},
 {'label': 'neutral', 'score': 0.932409942150116},
 {'label': 'neutral', 'score': 0.6844247579574585},
 {'label': 'positive', 'score': 0.943336546421051},
 {'label': 'positive', 'score': 0.9523476958274841},
 {'label': 'negative', 'score': 0.934123694896698},
 {'label': 'neutral', 'score': 0.9401276707649231},
 {'label': 'negative', 'score': 0.8472444415092468},
 {'label': 'positive', 'score': 0.9772746562957764}]

# **Test results and observation:**
| #  | Test Comment                                                        | Expected Sentiment      | Model Prediction |  Score | Result         | Observation                                                                                                                                      |
| -- | ------------------------------------------------------------------- | ----------------------- | ---------------- | -----: | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------ |
| 1  | You are doing great work.                                           | Positive                | **Positive**     | 97.12% | ✅ Correct      | Clear positive statement; model identified it confidently.                                                                                       |
| 2  | Well I don't trust this.                                            | Negative                | **Negative**     | 87.67% | ✅ Correct      | Expression of distrust was correctly identified as negative.                                                                                     |
| 3  | Does this organization work on the charity given by people?         | Neutral                 | **Neutral**      | 93.24% | ✅ Correct      | Informational question with no strong emotional tone.                                                                                            |
| 4  | Yeah I can see how much work is done.                               | Ambiguous / Challenging | **Neutral**      | 68.44% | ⚠️ Challenging | The meaning can depend on context or sarcasm, so neutral is a reasonable interpretation.                                                         |
| 5  | I'm happy this type of people exist who work for other people.      | Positive                | **Positive**     | 94.34% | ✅ Correct      | Clearly expresses appreciation and happiness.                                                                                                    |
| 6  | Thank you for supporting our community and helping those in need.   | Positive                | **Positive**     | 95.23% | ✅ Correct      | Gratitude and appreciation were correctly detected as positive.                                                                                  |
| 7  | I am disappointed with how this organization handled the donations. | Negative                | **Negative**     | 93.41% | ✅ Correct      | Explicit disappointment was correctly classified as negative.                                                                                    |
| 8  | Where can I find information about your next campaign?              | Neutral                 | **Neutral**      | 94.01% | ✅ Correct      | Simple information-seeking question; no strong sentiment.                                                                                        |
| 9  | Another campaign, but nothing ever seems to change.                 | Negative / Challenging  | **Neutral**      | 84.74% | ❌ Mismatch     | The statement implies dissatisfaction, but the model classified it as neutral. This may indicate difficulty with implicit negativity or sarcasm. |
| 10 | The volunteers have done an amazing job helping local families.     | Positive                | **Positive**     | 97.73% | ✅ Correct      | Strong positive language was correctly identified.                                                                                               |


# **Conclusion**
The hugging Face Transformer model is tested successfully for sentiment analysis. The model was tested on  positive, negative, neutral and challenging test samples, mostly sentiments were classified correctly and model provided the result with correct prediction and score, although at test no. 9 it didn't match the expected sentiment.